### DISCUSSION TITLE
Iterating over list of objects.

### DISCUSSION THREAD

**Discussion by eyepatch61** [nlohmann/json#2243] [Q&A] [answered]
Hi I have been trying to iterate over a list of objects using this library in a cpp file. Structure is like this :

`list : [`
`    {"foo": 4, "bar" : 6},`
`    {"foo": 8, "bar" : 6, "foo2" : 5},`
`    {"foo": 4, "bar" : 9}`
`]`

So for this my code would go like :
`for (json::iterator listObject = list.begin(); listObject != list.end(); listObject++) {`
`for (json::iterator option = listObject.begin(); option != listObject.end(); option++){`
`//implementation`
`}`
`}`

I thought I could use iterator for an object as well but I am facing an error for inner "for loop" which says `error: 'using iterator = nlohmann::basic_json<>::iter_impl<nlohmann::basic_json<> > {aka class nlohmann::detail::iter_impl<nlohmann::basic_json<> >}' has no member named 'begin'; did you mean 'set_begin'?
`

Could anyone tell me where I am doing wrong.

Thanks.

**Comment by eyepatch61** [comment]
This example doesn't seem to be shown in documentation as well. I'd like documentation to explain with examples with most use cases. Would help new developers to use this library.

**Comment by nlohmann** [comment, marked as answer]
The error message is not too much related to the library, but you could have the same issue if you would iterate a `std::vector<std::vector<int>>`: iterators need to be dereferenced to access the underlying value. So instead of `listObject.begin()`, try `listObject->begin()`.

Note you can simplify your code with range for (i.e., using `for (auto& val: list)`) and [`items()`](https://nlohmann.github.io/json/doxygen/classnlohmann_1_1basic__json_a5961446010dfc494e0c247b4e9026977.html#a5961446010dfc494e0c247b4e9026977).

**Reply by eyepatch61** [reply]
Thanks for the answer. it is working. Although I have been using this library. But never had to dereference while using objects/ objects of objects. But do I need to do it in the case of list of objects. Does it mean it loop over list iterates over the pointers to objects ?

**Reply by nlohmann** [reply]
You need to dereference an iterator to access its value. Like this:

```cpp
#include <vector>
#include <iostream>

int main()
{
    std::vector<std::vector<int>> nested;
    nested.push_back({{1,2,3}});
    nested.push_back({{2,4,6}});

    for (auto it1 = nested.begin(); it1 != nested.end(); ++it1)
    {
        for (auto it2 = it1->begin(); it2 != it1->end(); ++it2) // note calling it1->
        {
            std::cout << *it2 << " "; // note using *it2
        }
        std::cout << std::endl;
    }
}
```

This is basic iterator handling and not specific to this library.

**Reply by eyepatch61** [reply]
Yes It already worked. I was just asking why does it become a pointer in the second loop. I didn't face this earlier using this library.

**Reply by nlohmann** [reply]
It’s not a pointer - it’s an iterator.
